# 06 E2 Project

* ist11135321 Afonso Costa (33,3%)
  
* ist11146982 Lourenço Tavares (33,3%)
  
* ist11150863 Rafael Caratão (33,3%)

Prof. Diogo Ferreira and Prof. José Farinha

Lab Shift number: BD25610L-07

## Introdução

Após o diagrama Entidade-Associação para a base de dados “Zoo” ter sido apresentado ao cliente numa primeira entrega, seguiram-se várias iterações para alinhar o desenho da base de dados de acordo com os requisitos do cliente, tendo-se finalmente convergido no esquema SQL apresentado abaixo. 

Pretende-se agora implementar esta base de dados como parte de um sistema de informação que permita a gestão e análise da venda de bilhetes e da popularidade dos vários recintos e animais. A popularidade é um requisito novo, que advém de uma campanha do zoo: cada portador de *bilhete* tem direito a votar no seu *recinto* favorito, o que, na base de dados, incrementa *votos* do *recinto* em 1 e muda *votou* para TRUE no *bilhete*.

`Nota` Optou-se por um sistema de informação separado para a gestão dos funcionários do zoo, que não será abordado no presente trabalho.

## PART I – Base de Dados

#### 0. Criação da Base de Dados

Crie a base de dados `Zoo` no PostgreSQL e execute (nesta base de dados) os comandos para a implementação do respectivo esquema.

In [1]:
%load_ext sql
%config SqlMagic.displaycon = 0
%config SqlMagic.displaylimit = 100
%config SqlMagic.feedback = 0
%sql postgresql+psycopg://app:app@postgres/app --alias zoo

In [2]:
%%sql zoo

DROP MATERIALIZED VIEW IF EXISTS vendas_zoo;

++
||
++
++

In [3]:
%%sql zoo
DROP TABLE IF EXISTS acesso;
DROP TABLE IF EXISTS bilhete;
DROP TABLE IF EXISTS venda;
DROP TABLE IF EXISTS animal;
DROP TABLE IF EXISTS especie;
DROP TABLE IF EXISTS recinto;
DROP TABLE IF EXISTS zona;
DROP TYPE IF EXISTS cat;
DROP TYPE IF EXISTS cnt;

CREATE TYPE cat AS ENUM(
  'Aves',
  'Carnívoros',
  'Herbívoros',
  'Mamíferos Marinhos',
  'Primatas',
  'Repteis'
);

CREATE TYPE cnt AS ENUM(
  'África',
  'América',
  'Asia',
  'Austrália',
  'Europa'
);

CREATE TABLE zona (
  id_zona SERIAL PRIMARY KEY,
  categoria cat,
  continente cnt,
  preco NUMERIC(4, 2) NOT NULL,
  CONSTRAINT chave_zona UNIQUE (categoria, continente)
);

CREATE TABLE recinto (
  id_recinto SERIAL PRIMARY KEY,
  id_zona INTEGER NOT NULL REFERENCES zona,
  votos INTEGER
);

CREATE TABLE especie (
  nome_cientifico VARCHAR(100) PRIMARY KEY CHECK (nome_cientifico ~ '^[A-Z][a-z]+ [a-z]+$'),
  nome_comum VARCHAR(100) NOT NULL UNIQUE,
  categoria cat NOT NULL,
  continente cnt NOT NULL
);

CREATE TABLE animal (
  id_animal SERIAL PRIMARY KEY,
  nome VARCHAR(80) NOT NULL,
  nome_cientifico VARCHAR(100) NOT NULL REFERENCES especie,
  id_recinto INTEGER NOT NULL REFERENCES recinto,
  data_nasc DATE,
  CONSTRAINT chave_animal UNIQUE (nome, nome_cientifico)
);

CREATE TABLE venda (
  no_venda SERIAL PRIMARY KEY,
  data_hora TIMESTAMP NOT NULL,
  nif_cliente CHAR(9) CHECK (nif_cliente ~ '^[0-9]{9}$')
);

CREATE TABLE bilhete (
  bid SERIAL PRIMARY KEY,
  desconto NUMERIC(4, 2) NOT NULL DEFAULT 0.00,
  votou BOOLEAN NOT NULL DEFAULT FALSE,
  no_venda INTEGER NOT NULL REFERENCES venda
);

CREATE TABLE acesso (
  bid INTEGER NOT NULL REFERENCES bilhete,
  id_zona INTEGER NOT NULL REFERENCES zona,
  PRIMARY KEY (bid, id_zona)
);

++
||
++
++

Verifique que todas as tabelas foram criadas na base de dados com o seguinte comando:

In [4]:
%sqlcmd tables

Name
zona
recinto
especie
animal
venda
bilhete
acesso


#### 1. Restrições de Integridade [4 valores]

Recorrendo a `Triggers apenas quando estritamente necessário`, implemente na base de dados “Zoo” as seguintes restrições de integridade:


1. `RI-1` Em zona, a categoria e o continente não podem ser simultaneamente `NULL`
* uma zona é sempre dedicada a uma categoria de animais, a um continente, ou a ambos.

In [5]:
%%sql

ALTER TABLE zona
ADD CONSTRAINT ri1
CHECK (
    categoria IS NOT NULL 
    OR continente IS NOT NULL
);

++
||
++
++

2. `RI-2` Um animal tem de estar alojado num recinto que esteja numa zona compatível
* se a zona tiver categoria definida, tem de corresponder à categoria da espécie do animal;
* se tiver continente definido, tem de corresponder ao continente da espécie do animal.

In [6]:
%%sql zoo

DROP TRIGGER IF EXISTS ri2 ON animal;
DROP FUNCTION IF EXISTS ri2();

CREATE OR REPLACE FUNCTION ri2()
RETURNS TRIGGER AS
$$
BEGIN
    IF NOT EXISTS (
        SELECT *
        FROM especie e, recinto r, zona z
        WHERE e.nome_cientifico = NEW.nome_cientifico
          AND r.id_recinto = NEW.id_recinto
          AND z.id_zona = r.id_zona
          AND (z.categoria IS NULL OR z.categoria = e.categoria)
          AND (z.continente IS NULL OR z.continente = e.continente)
    ) THEN
        RAISE EXCEPTION 'RI-2 violada: o animal não cumpre as condições de alojamento.';
    END IF;

    RETURN NEW;
END
$$ LANGUAGE plpgsql;

CREATE TRIGGER ri2
BEFORE INSERT OR UPDATE ON animal
FOR EACH ROW
EXECUTE FUNCTION ri2();

++
||
++
++

`RI-3` Todos os animais de uma espécie têm de estar alojados em recinto(s) de uma mesma zona
* `Nota` No entanto, podem existir várias zonas compatíveis com a espécie.

In [7]:
%%sql zoo

DROP TRIGGER IF EXISTS ri3_animal ON animal;
DROP TRIGGER IF EXISTS ri3_recinto ON recinto;
DROP FUNCTION IF EXISTS ri3_animal();
DROP FUNCTION IF EXISTS ri3_recinto();

-- ri3_animal - o inserir/alterar um animal na tabela animal
CREATE OR REPLACE FUNCTION ri3_animal()
RETURNS TRIGGER AS
$$
BEGIN
    IF EXISTS (
        SELECT *
        FROM animal a, recinto r1, recinto r2
        WHERE a.nome_cientifico = NEW.nome_cientifico
          AND a.id_recinto = r1.id_recinto
          AND NEW.id_recinto = r2.id_recinto
          AND r1.id_zona <> r2.id_zona
    ) THEN
        RAISE EXCEPTION 'RI-3 violada: já existem animais da mesma espécie em recintos de outra zona';
    END IF;

    RETURN NEW;
END
$$ LANGUAGE plpgsql;

CREATE TRIGGER ri3_animal
BEFORE INSERT OR UPDATE ON animal
FOR EACH ROW
EXECUTE FUNCTION ri3_animal();

-- ri3_recinto - ao alterar a zona de um recinto na tabela recinto.
CREATE OR REPLACE FUNCTION ri3_recinto()
RETURNS TRIGGER AS
$$
BEGIN
    IF EXISTS (
        SELECT *
        FROM animal a1, animal a2, recinto r2
        WHERE a1.id_recinto = NEW.id_recinto
          AND a1.nome_cientifico = a2.nome_cientifico
          AND a2.id_recinto = r2.id_recinto
          AND NEW.id_zona <> r2.id_zona
    ) THEN
        RAISE EXCEPTION 'RI-3 violada: não é possível alterar a zona deste recinto, porque isso colocaria animais da mesma espécie em zonas diferentes';
    END IF;

    RETURN NEW;
END
$$ LANGUAGE plpgsql;

CREATE TRIGGER ri3_recinto
BEFORE UPDATE OF id_zona ON recinto
FOR EACH ROW
EXECUTE FUNCTION ri3_recinto();

++
||
++
++

`RI-4` Após uma transação de venda
* uma venda tem de incluir pelo menos um bilhete com acesso a pelo menos uma zona do zoo.

In [8]:
%%sql zoo

DROP TRIGGER IF EXISTS ri4_venda ON venda;
DROP TRIGGER IF EXISTS ri4_bilhete ON bilhete;
DROP TRIGGER IF EXISTS ri4_acesso ON acesso;
DROP FUNCTION IF EXISTS ri4();
DROP FUNCTION IF EXISTS ri4_check(INTEGER);

CREATE OR REPLACE FUNCTION ri4_check(p_no_venda INTEGER)
RETURNS VOID AS
$$
BEGIN
    IF p_no_venda IS NOT NULL
       AND EXISTS (
            SELECT *
            FROM venda v
            WHERE v.no_venda = p_no_venda
       )
       AND NOT EXISTS (
            SELECT *
            FROM bilhete b, acesso a
            WHERE b.bid = a.bid
              AND b.no_venda = p_no_venda
       )
    THEN
        RAISE EXCEPTION 'RI-4 violada: venda inválida';
    END IF;
END
$$ LANGUAGE plpgsql;


CREATE OR REPLACE FUNCTION ri4()
RETURNS TRIGGER AS
$$
DECLARE
    v_no_venda INTEGER;
BEGIN
    IF TG_TABLE_NAME = 'venda' THEN
        IF TG_OP = 'DELETE' THEN
            PERFORM ri4_check(OLD.no_venda);
            RETURN OLD;
        ELSE
            PERFORM ri4_check(NEW.no_venda);
            RETURN NEW;
        END IF;
    END IF;

    IF TG_TABLE_NAME = 'bilhete' THEN
        IF TG_OP = 'DELETE' THEN
            PERFORM ri4_check(OLD.no_venda);
            RETURN OLD;
        ELSIF TG_OP = 'UPDATE' THEN
            PERFORM ri4_check(OLD.no_venda);
            PERFORM ri4_check(NEW.no_venda);
            RETURN NEW;
        ELSE
            PERFORM ri4_check(NEW.no_venda);
            RETURN NEW;
        END IF;
    END IF;

    IF TG_TABLE_NAME = 'acesso' THEN
        IF TG_OP = 'DELETE' THEN
            SELECT b.no_venda INTO v_no_venda
            FROM bilhete b
            WHERE b.bid = OLD.bid;

            PERFORM ri4_check(v_no_venda);
            RETURN OLD;

        ELSIF TG_OP = 'UPDATE' THEN
            SELECT b.no_venda INTO v_no_venda
            FROM bilhete b
            WHERE b.bid = OLD.bid;

            PERFORM ri4_check(v_no_venda);

            SELECT b.no_venda INTO v_no_venda
            FROM bilhete b
            WHERE b.bid = NEW.bid;

            PERFORM ri4_check(v_no_venda);
            RETURN NEW;

        ELSE
            SELECT b.no_venda INTO v_no_venda
            FROM bilhete b
            WHERE b.bid = NEW.bid;

            PERFORM ri4_check(v_no_venda);
            RETURN NEW;
        END IF;
    END IF;

    RETURN NULL;
END
$$ LANGUAGE plpgsql;


CREATE CONSTRAINT TRIGGER ri4_venda
AFTER INSERT OR UPDATE OR DELETE ON venda
DEFERRABLE INITIALLY DEFERRED
FOR EACH ROW
EXECUTE FUNCTION ri4();

CREATE CONSTRAINT TRIGGER ri4_bilhete
AFTER INSERT OR UPDATE OR DELETE ON bilhete
DEFERRABLE INITIALLY DEFERRED
FOR EACH ROW
EXECUTE FUNCTION ri4();

CREATE CONSTRAINT TRIGGER ri4_acesso
AFTER INSERT OR UPDATE OR DELETE ON acesso
DEFERRABLE INITIALLY DEFERRED
FOR EACH ROW
EXECUTE FUNCTION ri4();

++
||
++
++

#### 2. Preenchimento da Base de Dados [3 valores]

##### Critérios de Avaliação

* Cumprimento dos requisitos de cobertura  
* Resultado não vazio em todas as consultas do ponto 5

`Nota`: Caso não consiga implementar alguma(s) das restrições de integridade do ponto 1, deve ainda assim assegurar-se que o preenchimento da base dados as cumpre. 

Pode utilizar ferramentas de **IA generativo**, scripts ou qualquer outro meio para gerar os comandos INSERT. Preencha todas as tabelas da base de dados de forma consistente, após execução do ponto anterior para garantir que são respeitadas todas as restrições de integridade ali expressas, e com os seguintes **requisitos de cobertura** adicionais:

1. Têm de haver pelo menos 7 ***zonas***, das quais:
* O preço de acesso a cada zona deve estar entre 5€ e 30€.
* pelo menos 3 são dedicadas exclusivamente a uma *categoria* (sem *continente* preenchido);
* pelo menos 2 são dedicadas exclusivamente a um *continente* (sem *categoria* preenchida);
* pelo menos 2 têm *categoria* e *continente* preenchidos.

`Nota` As 2 ou mais zonas que têm *categoria* e *continente* preenchidos têm de partilhar entre todas elas um único valor de um desses dois atributos (sendo naturalmente o outro diferente, devido à restrição de chave candidata da tabela), e o valor desse atributo não pode ocorrer sozinho (i.e., qualquer zona que tenha o valor desse atributo não pode ter o outro atributo a NULL). O valor desse atributo corresponde à “especialidade” do zoo.  
* `Exemplo 1` Um zoo pode ter como especialidade herbívoros, e nesse caso terá de ter pelo menos 2 zonas com categoria “herbívoro” e continentes diferentes, não podendo ter nenhuma zona com categoria “herbívoro” sem continente definido. Terá adicionalmente pelo menos 3 zonas dedicadas a outras categorias que não “herbívoro” e 2 zonas dedicadas a continentes.  
* `Exemplo 2` Outro zoo pode ter como especialidade animais de África, e nesse caso terá de ter pelo menos 2 zonas com continente “África” e categorias diferentes, não podendo ter nenhuma zona com continente “África” sem categoria definida. Terá adicionalmente pelo menos 3 zonas dedicadas a categorias e 2 zonas dedicadas a continentes que não “África”.


In [9]:
%%sql zoo

INSERT INTO zona (categoria, continente, preco) VALUES
-- apenas categoria
('Carnívoros', NULL, 18.50),
('Primatas', NULL, 20.00),
('Repteis', NULL, 15.00),

-- apenas continente
(NULL, 'África', 12.00),
(NULL, 'Europa', 10.00),

-- categoria + continente (especialidade do zoo = Herbívoros)
('Herbívoros', 'América', 22.00),
('Herbívoros', 'Asia', 24.00),
('Herbívoros', 'Austrália', 29.00);


++
||
++
++

2. Cada *zona* tem de ter entre 10 e 30 ***recintos***, e cada recinto deve ter no mínimo 0.1% dos votos totais (ver 5. abaixo). Os votos devem refletir os *acessos* dos *bilhetes*:
* O somatório global dos *votos* de todos os ***recintos*** tem de ser igual ao número de ***bilhetes*** vendidos com *votou* TRUE; 
* O somatório dos *votos* de todos os ***recintos*** de uma ***zona*** tem de ser menor ou igual ao número de ***bilhetes*** vendidos com acesso a essa ***zona*** e *votou* TRUE.

`Nota`: Pode preencher os votos dos ***recintos*** já no INSERT destes, ou alternativamente após o preenchimento dos *bilhetes*, por meio de comandos de UPDATE.

In [10]:
%%sql
INSERT INTO recinto (id_zona, votos) VALUES

-- zona 1
(1,1),(1,1),(1,1),(1,1),(1,1),
(1,1),(1,1),(1,1),(1,1),(1,1),
(1,1),(1,1),(1,1),(1,1),(1,1),
(1,1),(1,1),(1,1),(1,1),(1,1),

-- zona 2
(2,1),(2,1),(2,1),(2,1),(2,1),
(2,1),(2,1),(2,1),(2,1),(2,1),
(2,1),(2,1),(2,1),(2,1),(2,1),
(2,1),(2,1),(2,1),(2,1),(2,1),

-- zona 3
(3,1),(3,1),(3,1),(3,1),(3,1),
(3,1),(3,1),(3,1),(3,1),(3,1),
(3,1),(3,1),(3,1),(3,1),(3,1),
(3,1),(3,1),(3,1),(3,1),(3,1),

-- zona 4
(4,1),(4,1),(4,1),(4,1),(4,1),
(4,1),(4,1),(4,1),(4,1),(4,1),
(4,1),(4,1),(4,1),(4,1),(4,1),
(4,1),(4,1),(4,1),(4,1),(4,1),

-- zona 5
(5,1),(5,1),(5,1),(5,1),(5,1),
(5,1),(5,1),(5,1),(5,1),(5,1),
(5,1),(5,1),(5,1),(5,1),(5,1),
(5,1),(5,1),(5,1),(5,1),(5,1),

-- zona 6
(6,1),(6,1),(6,1),(6,1),(6,1),
(6,1),(6,1),(6,1),(6,1),(6,1),
(6,1),(6,1),(6,1),(6,1),(6,1),
(6,1),(6,1),(6,1),(6,1),(6,1),

-- zona 7
(7,1),(7,1),(7,1),(7,1),(7,1),
(7,1),(7,1),(7,1),(7,1),(7,1),
(7,1),(7,1),(7,1),(7,1),(7,1),
(7,1),(7,1),(7,1),(7,1),(7,1),

-- zona 8
(8,1),(8,1),(8,1),(8,1),(8,1),
(8,1),(8,1),(8,1),(8,1),(8,1),
(8,1),(8,1),(8,1),(8,1),(8,1),
(8,1),(8,1),(8,1),(8,1),(8,1);

++
||
++
++

3. Têm de haver pelo menos 200 ***espécies*** reais cobrindo todas as *categorias* e todos os *continentes*.  

In [11]:
%%sql
INSERT INTO especie
(nome_cientifico, nome_comum, categoria, continente)
VALUES
-- ÁFRICA
('Panthera leo', 'Leão', 'Carnívoros', 'África'),
('Loxodonta africana', 'Elefante-africano', 'Herbívoros', 'África'),
('Gorilla gorilla', 'Gorila-ocidental', 'Primatas', 'África'),
('Python sebae', 'Pitão-africana', 'Repteis', 'África'),
('Acinonyx jubatus', 'Chita', 'Carnívoros', 'África'),
('Panthera pardus','Leopardo','Carnívoros','África'),
('Lycaon pictus','Cao-selvagem-africano','Carnívoros','África'),
('Syncerus caffer','Bufalo-africano','Herbívoros','África'),
('Hippopotamus amphibius','Hipopotamo','Herbívoros','África'),
('Papio anubis','Babuíno-oliva','Primatas','África'),
('Colobus guereza','Colobo-preto-branco','Primatas','África'),
('Struthio camelus','Avestruz','Aves','África'),
('Phoenicopterus roseus','Flamingo','Aves','África'),
('Varanus niloticus','Varano-nilo','Repteis','África'),
('Crocodylus niloticus','Crocodilo-nilo','Repteis','África'),

-- AMÉRICA
('Bison bison', 'Bisão-americano', 'Herbívoros', 'América'),
('Puma concolor', 'Puma', 'Carnívoros', 'América'),
('Alouatta palliata', 'Bugio', 'Primatas', 'América'),
('Caiman crocodilus', 'Jacaré-de-óculos', 'Repteis', 'América'),
('Odocoileus virginianus', 'Veado-cauda-branca', 'Herbívoros', 'América'), --DUVIDAS " não colocamos as 200 especies para ser mais facil testar"
('Ursus americanus','Urso-negro','Carnívoros','América'),
('Procyon lotor','Guaxinim','Carnívoros','América'),
('Odocoileus hemionus','Veado-mula','Herbívoros','América'),
('Tapirus terrestris','Anta','Herbívoros','América'),
('Ateles geoffroyi','Macaco-aranha','Primatas','América'),
('Cebus capucinus','Macaco-prego','Primatas','América'),
('Ara macao','Arara-vermelha','Aves','América'),
('Haliaeetus leucocephalus','Aguia-careca','Aves','América'),
('Iguana iguana','Iguana-verde','Repteis','América'),
('Chelonoidis nigra','Tartaruga-gigante','Repteis','América'),


-- ASIA
('Panthera tigris', 'Tigre', 'Carnívoros', 'Asia'),
('Elephas maximus', 'Elefante-asiático', 'Herbívoros', 'Asia'),
('Macaca fuscata', 'Macaco-japonês', 'Primatas', 'Asia'),
('Python reticulatus', 'Pitão-reticulada', 'Repteis', 'Asia'),
('Ailuropoda melanoleuca', 'Panda-gigante', 'Herbívoros', 'Asia'),
('Panthera uncia','Leopardo-neves','Carnívoros','Asia'),
('Cuon alpinus','Cao-selvagem-asiatico','Carnívoros','Asia'),
('Bos gaurus','Gauro','Herbívoros','Asia'),
('Rhinoceros unicornis','Rinoceronte-indiano','Herbívoros','Asia'),
('Trachypithecus francoisi','Langur-francois','Primatas','Asia'),
('Nasalis larvatus','Macaco-narigudo','Primatas','Asia'),
('Pavo cristatus','Pavao-indiano','Aves','Asia'),
('Grus japonensis','Grou-japones','Aves','Asia'),
('Varanus salvator','Varano-asiatico','Repteis','Asia'),
('Gavialis gangeticus','Gavial','Repteis','Asia'),

-- EUROPA
('Canis lupus', 'Lobo-cinzento', 'Carnívoros', 'Europa'),
('Cervus elaphus', 'Veado-vermelho', 'Herbívoros', 'Europa'),
('Lacerta bilineata', 'Lagarto-verde', 'Repteis', 'Europa'),
('Ursus arctos', 'Urso-pardo', 'Carnívoros', 'Europa'),
('Lynx lynx','Lince-euroasiatico','Carnívoros','Europa'),
('Vulpes vulpes','Raposa-vermelha','Carnívoros','Europa'),
('Capreolus capreolus','Corco','Herbívoros','Europa'),
('Rupicapra rupicapra','Camurca','Herbívoros','Europa'),
('Macaca sylvanus','Macaco-barbaria','Primatas','Europa'),
('Delphinus delphis','Golfinho-comum','Mamíferos Marinhos','Europa'),
('Cygnus olor','Cisne-branco','Aves','Europa'),
('Bubo bubo','Bufo-real','Aves','Europa'),
('Vipera berus','Vibora-europeia','Repteis','Europa'),
('Testudo hermanni','Tartaruga-hermann','Repteis','Europa'),

-- AUSTRALIA
('Macropus rufus','Canguru-vermelho','Herbívoros','Austrália'),
('Phascolarctos cinereus','Coala','Herbívoros','Austrália'),
('Vombatus ursinus','Vombate','Herbívoros','Austrália'),
('Sarcophilus harrisii','Diabo-tasmania','Carnívoros','Austrália'),
('Dasyurus viverrinus','Quol-oriental','Carnívoros','Austrália'),
('Ornithorhynchus anatinus','Ornitorrinco','Mamíferos Marinhos','Austrália'),
('Tursiops truncatus','Golfinho-roaz','Mamíferos Marinhos','Austrália'),
('Dromaius novaehollandiae','Emu','Aves','Austrália'),
('Cacatua galerita','Cacatua-amarela','Aves','Austrália'),
('Morelia spilota','Piton-tapete','Repteis','Austrália');

++
||
++
++

4. O número de ***animais*** de cada *espécie* deve ser entre 1 e 12, com média entre 2 e 3 animais
* Um animal só pode estar num *recinto* sozinho se há 3 ou menos ***animais*** dessa *espécie*.
* Têm de haver no zoo alguns *recintos* com:
    * apenas um ***animal***;
    * vários ***animais*** de uma só *espécie*;
    * vários ***animais*** de várias *espécies*.

In [12]:
%%sql

INSERT INTO animal (nome, nome_cientifico, id_recinto, data_nasc)
VALUES

-- Zona 1 (Carnívoros)
('Leo1','Panthera leo',1,'2020-01-01'),
('Leo2','Panthera leo',2,'2021-01-01'),
('Chi1','Acinonyx jubatus',3,'2019-01-01'),
('Chi2','Acinonyx jubatus',4,'2020-01-01'),
('Pum1','Puma concolor',5,'2020-01-01'),
('Pum2','Puma concolor',6,'2021-01-01'),
('Tig1','Panthera tigris',7,'2020-01-01'),
('Tig2','Panthera tigris',8,'2021-01-01'),
('Tig3','Panthera tigris',9,'2022-01-01'),
('Leop1','Panthera pardus',10,'2020-05-01'),
('Leop2','Panthera pardus',11,'2021-05-01'),
('Lin1','Lynx lynx',12,'2020-01-01'),
('Lin2','Lynx lynx',13,'2021-01-01'),
('Rap1','Vulpes vulpes',14,'2020-01-01'),
('Rap2','Vulpes vulpes',15,'2021-01-01'),

-- Zona 2 (Primatas)
('Gor1','Gorilla gorilla',21,'2017-05-01'),
('Gor2','Gorilla gorilla',22,'2018-06-01'),
('Bug1','Alouatta palliata',23,'2020-01-01'),
('Bug2','Alouatta palliata',24,'2021-01-01'),
('Mac1','Macaca fuscata',25,'2020-01-01'),
('Mac2','Macaca fuscata',26,'2021-01-01'),
('Mac3','Macaca fuscata',27,'2022-01-01'),
('Bab1','Papio anubis',28,'2020-01-01'),
('Bab2','Papio anubis',29,'2021-01-01'),
('Col1','Colobus guereza',30,'2020-01-01'),
('Col2','Colobus guereza',31,'2021-01-01'),
('Lan1','Trachypithecus francoisi',32,'2020-01-01'),
('Lan2','Trachypithecus francoisi',33,'2021-01-01'),
('Nar1','Nasalis larvatus',34,'2020-01-01'),
('Nar2','Nasalis larvatus',35,'2021-01-01'),

-- Zona 3 (Repteis)
('Pit1','Python sebae',41,'2022-01-01'),
('Pit2','Python sebae',42,'2023-01-01'),
('Jac1','Caiman crocodilus',43,'2020-01-01'),
('Jac2','Caiman crocodilus',44,'2021-01-01'),
('PitR1','Python reticulatus',45,'2020-01-01'),
('PitR2','Python reticulatus',46,'2021-01-01'),
('PitR3','Python reticulatus',47,'2022-01-01'),
('Cro1','Crocodylus niloticus',48,'2020-01-01'),
('Cro2','Crocodylus niloticus',49,'2021-01-01'),
('Gav1','Gavialis gangeticus',50,'2020-01-01'),
('Gav2','Gavialis gangeticus',51,'2021-01-01'),

-- Zona 4 (África)
('Ele1','Loxodonta africana',61,'2018-03-01'),
('Ele2','Loxodonta africana',62,'2019-04-01'),
('Buf1','Syncerus caffer',63,'2019-01-01'),
('Buf2','Syncerus caffer',64,'2020-01-01'),
('Hip1','Hippopotamus amphibius',65,'2018-01-01'),
('Hip2','Hippopotamus amphibius',66,'2019-01-01'),
('Ave1','Struthio camelus',67,'2020-01-01'),
('Ave2','Struthio camelus',68,'2021-01-01'),
('Fla1','Phoenicopterus roseus',69,'2020-01-01'),
('Fla2','Phoenicopterus roseus',70,'2021-01-01'),

-- Zona 5 (Europa)
('Lob1','Canis lupus',81,'2020-01-01'),
('Lob2','Canis lupus',82,'2021-01-01'),
('Lob3','Canis lupus',83,'2022-01-01'),
('Cer1','Cervus elaphus',84,'2020-01-01'),
('Cer2','Cervus elaphus',85,'2021-01-01'),
('Cer3','Cervus elaphus',86,'2022-01-01'),
('Gib1','Macaca sylvanus',87,'2020-01-01'),
('Gib2','Macaca sylvanus',88,'2021-01-01'),
('Gib3','Macaca sylvanus',89,'2022-01-01'),
('Lag1','Lacerta bilineata',90,'2020-01-01'),
('Lag2','Lacerta bilineata',91,'2021-01-01'),
('Lag3','Lacerta bilineata',92,'2022-01-01'),
('Urs1','Ursus arctos',93,'2020-01-01'),
('Urs2','Ursus arctos',94,'2021-01-01'),
('Urs3','Ursus arctos',95,'2022-01-01'),
('Cor1','Capreolus capreolus',96,'2020-01-01'),
('Cor2','Capreolus capreolus',97,'2021-01-01'),
('Cam1','Rupicapra rupicapra',98,'2020-01-01'),
('Cam2','Rupicapra rupicapra',99,'2021-01-01'),
('Cis1','Cygnus olor',100,'2020-01-01'),

-- Zona 6 (Herbívoros + América)
('Bis1','Bison bison',101,'2018-01-01'),
('Bis2','Bison bison',102,'2019-01-01'),
('Vea1','Odocoileus virginianus',103,'2020-01-01'),
('Vea2','Odocoileus virginianus',104,'2021-01-01'),
('Tap1','Tapirus terrestris',105,'2020-01-01'),
('Tap2','Tapirus terrestris',106,'2021-01-01'),

-- Zona 7 (Herbívoros + Asia)
('EleA1','Elephas maximus',121,'2020-01-01'),
('EleA2','Elephas maximus',122,'2021-01-01'),
('EleA3','Elephas maximus',123,'2022-01-01'),
('Pan1','Ailuropoda melanoleuca',124,'2020-01-01'),
('Pan2','Ailuropoda melanoleuca',125,'2021-01-01'),
('Pan3','Ailuropoda melanoleuca',126,'2022-01-01'),
('Uni1','Rhinoceros unicornis',127,'2020-01-01'),
('Uni2','Rhinoceros unicornis',128,'2021-01-01'),
('Gau1','Bos gaurus',129,'2020-01-01'),
('Gau2','Bos gaurus',130,'2021-01-01'),

-- Zona 8 (Herbívoros + Austrália)
('Can1','Macropus rufus',141,'2020-01-01'),
('Can2','Macropus rufus',142,'2021-01-01'),
('Coa1','Phascolarctos cinereus',143,'2020-01-01'),
('Coa2','Phascolarctos cinereus',144,'2021-01-01'),
('Vom1','Vombatus ursinus',145,'2020-01-01'),
('Vom2','Vombatus ursinus',146,'2021-01-01');

++
||
++
++

In [13]:
%%sql

DROP TRIGGER ri4_venda ON venda;
DROP TRIGGER ri4_bilhete on bilhete;
DROP TRIGGER ri4_acesso ON acesso;

++
||
++
++

5. Têm de haver ***vendas*** de ***bilhetes*** para todos os dias de 2026 até 11 de Junho (i.e., `[2026-01-01 - 2026-06-11]`), tais que:
* Haja pelo menos 1000 ***bilhetes*** nos dias de semana;
* Haja pelo menos 4000 ***bilhetes*** nos fins de semana;
* Cerca de metade dos ***bilhetes*** por dia sejam para crianças ou séniores, tendo 50% de *desconto*, e os restantes não têm *desconto*;
* Pelo menos 2% dos ***bilhetes*** de cada dia tenham `Acesso Total` (i.e., ***acesso*** a todas as *zonas*);
* Todas as ***zonas*** estejam em pelo menos 10 ***bilhetes*** por dia;
* Cada ***bilhete*** tenha ***acesso*** a 3 ou mais ***zonas***, e cada ***zona*** tem de figurar em pelo menos 25% dos ***bilhetes***;
* Tem de haver ***bilhetes*** para todas as combinações de 3 ou mais ***zonas*** (mas não necessariamente todos os dias);
* Pelo menos 75% dos ***bilhetes*** têm *votou* a TRUE.

`Nota` A implementação da RI-4 no ponto anterior obriga a que a inserção nestas três tabelas seja encapsulada numa transação. 

In [14]:
%%sql zoo

-- =====================================================
-- VENDAS (1 venda por dia)
-- =====================================================

INSERT INTO venda (data_hora, nif_cliente)
SELECT
    dia + interval '12 hours',
    LPAD(((random()*999999999)::bigint)::text,9,'0')
FROM generate_series(
    '2026-01-01'::timestamp,
    '2026-06-11'::timestamp,
    interval '1 day'
) AS dia;

-- =====================================================
-- BILHETES
-- ~7500 no total
-- =====================================================

INSERT INTO bilhete (desconto, votou, no_venda)
SELECT
    CASE
        WHEN random() < 0.5 THEN 0.50
        ELSE 0.00
    END,
    random() < 0.75,
    v.no_venda
FROM venda v,
LATERAL generate_series(
    1,
    CASE
        WHEN EXTRACT(ISODOW FROM v.data_hora) IN (6,7)
            THEN 4000
        ELSE 1000
    END
);

-- =====================================================
-- TODOS OS BILHETES TÊM PELO MENOS 3 ZONAS
-- =====================================================

INSERT INTO acesso (bid, id_zona)
SELECT bid, 1
FROM bilhete;

INSERT INTO acesso (bid, id_zona)
SELECT bid, 2
FROM bilhete;

INSERT INTO acesso (bid, id_zona)
SELECT bid, 3
FROM bilhete;

-- =====================================================
-- ZONA 4 EM ~50% DOS BILHETES
-- =====================================================

INSERT INTO acesso (bid, id_zona)
SELECT bid, 4
FROM bilhete
WHERE bid % 2 = 0;

-- =====================================================
-- ZONA 5 EM ~50% DOS BILHETES
-- =====================================================

INSERT INTO acesso (bid, id_zona)
SELECT bid, 5
FROM bilhete
WHERE bid % 2 = 1;

-- =====================================================
-- ZONA 6 EM ~33% DOS BILHETES
-- =====================================================

INSERT INTO acesso (bid, id_zona)
SELECT bid, 6
FROM bilhete
WHERE bid % 3 = 0;

-- =====================================================
-- ZONA 7 EM ~33% DOS BILHETES
-- =====================================================

INSERT INTO acesso (bid, id_zona)
SELECT bid, 7
FROM bilhete
WHERE bid % 3 = 1;

-- =====================================================
-- ~2% COM ACESSO TOTAL
-- =====================================================

INSERT INTO acesso (bid, id_zona)
SELECT bid, 4
FROM bilhete
WHERE bid % 50 = 0
ON CONFLICT DO NOTHING;

INSERT INTO acesso (bid, id_zona)
SELECT bid, 5
FROM bilhete
WHERE bid % 50 = 0
ON CONFLICT DO NOTHING;

INSERT INTO acesso (bid, id_zona)
SELECT bid, 6
FROM bilhete
WHERE bid % 50 = 0
ON CONFLICT DO NOTHING;

INSERT INTO acesso (bid, id_zona)
SELECT bid, 7
FROM bilhete
WHERE bid % 50 = 0
ON CONFLICT DO NOTHING;


++
||
++
++

## PART II – Sistema de Informação e Aplicação

#### 3. Desenvolvimento de Aplicação [4 valores]

##### Critérios de Avaliação

* Funcionalidade dos *endpoints*
* Uso correto de transações
* Prevenção de injeção de SQL
* Uso correto de métodos e códigos de erro HTTP

`Nota` Todo o código da aplicação deve ser incluído na entrega do projeto.

`Nota` A aplicação deve ainda estar disponível *online* para demonstração durante a discussão oral, no ambiente de desenvolvimento Docker providenciado para a disciplina.

Crie um protótipo de RESTful *web service* para venda de bilhetes e votação nos recintos por acesso programático à base de dados ‘zoo’ através de uma API que devolve respostas em JSON, análoga à demonstrada no Lab10. A API deve implementar os seguintes *endpoints REST*:

| Endpoint | Descrição |
| ----- | ----- |
| /zona/\<zona\>/ | OUTPUT: lista de todos os ***recintos*** da \<zona\>, contendo o *número* do ***recinto*** e as ***espécies*** nele contidas, indicando, para cada ***espécie***, os *nomes científico* e *comum*, e o número de ***animais*** da ***espécie*** que estão no ***recinto***. |
| /recinto/\<recinto\>/voto/\<bilhete\>/ | Assinala o voto do \<bilhete\> no \<recinto\>, atualizando as tabelas ***bilhete*** (marcando *votou* TRUE) e ***recinto*** (incrementando os *votos*). Devolve mensagem de erro informativa e diferenciada (e não assinala o voto) se o \<bilhete\> já tinha *votou* \= TRUE ou se o \<recinto\> não está contido numa zona a que o \<bilhete\> tinha acesso. |
| /venda/ | Executa  uma venda de um ou mais bilhetes, ***populando*** as tabelas ***venda***, ***bilhete*** e ***acesso***. INPUT: NIF do cliente \[opcional\] mais lista de bilhetes, em que cada bilhete consiste numa lista de zonas de acesso, e um desconto \[opcional\]. OUTPUT: O preço total da venda, e a lista dos bilhetes da venda com o número e preço de cada um.  |

A solução deve prezar pela segurança, **prevenindo** ataques por **injeção de** **SQL**, e garantindo que as **operações de escrita** estão encapsuladas em **transações** que cumprem os princípios ACID.

0. Itemize quais as estratégias que utilizou:

* Estratégia 1: A aplicação foi implementada como uma web API RESTful, em que a interação com a base de dados fica contida nos endpoints da API. As respostas são devolvidas em JSON e foram usados métodos HTTP adequados: GET para consultas e POST para operações que criam ou alteram dados.

* Estratégia 2: Para prevenir injeção de SQL, todas as interrogações foram escritas com parâmetros do Psycopg, usando placeholders, em vez de construir comandos SQL por concatenação de strings com valores recebidos do utilizador.

* Estratégia 3: As operações de escrita que envolvem várias tabelas foram executadas dentro de transações. Assim, no registo de uma venda, as inserções em venda, bilhete e acesso são tratadas como uma unidade lógica de trabalho; no voto, a alteração do bilhete e do recinto também é feita de forma atómica.

* Estratégia 4: No endpoint de voto foram usados SELECT ... FOR UPDATE para bloquear as linhas relevantes durante a transação. Esta estratégia evita problemas de concorrência, como dois pedidos tentarem usar o mesmo bilhete ao mesmo tempo.

1. Copie para a célula abaixo a função do app.py zona_index

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route("/zona/<int:zona>/", methods=("GET",))
def zona_index(zona):
    with pool.connection() as conn:
        with conn.cursor() as cur:
            zona_existe = cur.execute(
                """
                SELECT id_zona
                FROM zona
                WHERE id_zona = %(zona)s;
                """,
                {"zona": zona},
            ).fetchone()

            if zona_existe is None:
                return jsonify({"erro": "Zona inexistente."}), 404

            rows = cur.execute(
                """
                SELECT
                    r.id_recinto,
                    e.nome_cientifico,
                    e.nome_comum,
                    COUNT(a.id_animal) AS numero_animais
                FROM recinto r
                LEFT JOIN animal a
                    ON a.id_recinto = r.id_recinto
                LEFT JOIN especie e
                    ON e.nome_cientifico = a.nome_cientifico
                WHERE r.id_zona = %(zona)s
                GROUP BY
                    r.id_recinto,
                    e.nome_cientifico,
                    e.nome_comum
                ORDER BY
                    r.id_recinto,
                    e.nome_cientifico;
                """,
                {"zona": zona},
            ).fetchall()

    recintos = {}

    for row in rows:
        id_recinto = row.id_recinto

        if id_recinto not in recintos:
            recintos[id_recinto] = {
                "id_recinto": id_recinto,
                "especies": [],
            }

        if row.nome_cientifico is not None:
            recintos[id_recinto]["especies"].append(
                {
                    "nome_cientifico": row.nome_cientifico,
                    "nome_comum": row.nome_comum,
                    "numero_animais": row.numero_animais,
                }
            )

    return jsonify(list(recintos.values())), 200
```

2. Copie para a célula abaixo a função do app.py recinto_voto_save

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route("/recinto/<int:recinto>/voto/<int:bilhete>/", methods=("POST",))
def recinto_voto_save(recinto, bilhete):
    with pool.connection() as conn:
        with conn.transaction():
            with conn.cursor() as cur:
                bilhete_row = cur.execute(
                    """
                    SELECT bid, votou
                    FROM bilhete
                    WHERE bid = %(bilhete)s
                    FOR UPDATE;
                    """,
                    {"bilhete": bilhete},
                ).fetchone()

                if bilhete_row is None:
                    return jsonify({"erro": "Bilhete inexistente."}), 404

                if bilhete_row.votou:
                    return jsonify({"erro": "O bilhete já tinha votado."}), 409

                recinto_row = cur.execute(
                    """
                    SELECT id_recinto, id_zona
                    FROM recinto
                    WHERE id_recinto = %(recinto)s
                    FOR UPDATE;
                    """,
                    {"recinto": recinto},
                ).fetchone()

                if recinto_row is None:
                    return jsonify({"erro": "Recinto inexistente."}), 404

                tem_acesso = cur.execute(
                    """
                    SELECT 1
                    FROM acesso
                    WHERE bid = %(bilhete)s
                      AND id_zona = %(zona)s;
                    """,
                    {
                        "bilhete": bilhete,
                        "zona": recinto_row.id_zona,
                    },
                ).fetchone()

                if tem_acesso is None:
                    return jsonify(
                        {"erro": "O bilhete não tem acesso à zona deste recinto."}
                    ), 403

                cur.execute(
                    """
                    UPDATE bilhete
                    SET votou = TRUE
                    WHERE bid = %(bilhete)s;
                    """,
                    {"bilhete": bilhete},
                )

                cur.execute(
                    """
                    UPDATE recinto
                    SET votos = COALESCE(votos, 0) + 1
                    WHERE id_recinto = %(recinto)s;
                    """,
                    {"recinto": recinto},
                )

    return jsonify(
        {
            "mensagem": "Voto registado com sucesso.",
            "recinto": recinto,
            "bilhete": bilhete,
        }
    ), 200
```

3. Copie para a célula abaixo a função do app.py venda_save

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route("/venda/", methods=("POST",))
def venda_save():
    data = request.get_json(silent=True)

    if data is None:
        return jsonify({"erro": "O pedido tem de conter JSON."}), 400

    nif_cliente = data.get("nif_cliente")

    if nif_cliente == "":
        nif_cliente = None

    if nif_cliente is not None:
        if (
            not isinstance(nif_cliente, str)
            or len(nif_cliente) != 9
            or not nif_cliente.isdigit()
        ):
            return jsonify({"erro": "O NIF tem de ter 9 dígitos."}), 400

    bilhetes = data.get("bilhetes")

    if not isinstance(bilhetes, list) or len(bilhetes) == 0:
        return jsonify({"erro": "A venda tem de incluir pelo menos um bilhete."}), 400

    bilhetes_preparados = []

    for i, bilhete_data in enumerate(bilhetes, start=1):
        if not isinstance(bilhete_data, dict):
            return jsonify({"erro": f"O bilhete {i} é inválido."}), 400

        zonas = bilhete_data.get("zonas")

        if not isinstance(zonas, list) or len(zonas) == 0:
            return jsonify(
                {"erro": f"O bilhete {i} tem de ter pelo menos uma zona."}
            ), 400

        try:
            zonas = [int(zona) for zona in zonas]
        except (TypeError, ValueError):
            return jsonify(
                {"erro": f"As zonas do bilhete {i} têm de ser números inteiros."}
            ), 400

        if len(set(zonas)) != len(zonas):
            return jsonify({"erro": f"O bilhete {i} tem zonas repetidas."}), 400

        desconto = bilhete_data.get("desconto", 0)

        try:
            desconto = Decimal(str(desconto))
        except InvalidOperation:
            return jsonify({"erro": f"O desconto do bilhete {i} é inválido."}), 400

        if desconto < Decimal("0") or desconto > Decimal("1"):
            return jsonify(
                {"erro": f"O desconto do bilhete {i} tem de estar entre 0 e 1."}
            ), 400

        bilhetes_preparados.append(
            {
                "zonas": zonas,
                "desconto": desconto,
            }
        )

    resultado_bilhetes = []
    preco_total = Decimal("0.00")

    try:
        with pool.connection() as conn:
            with conn.transaction():
                with conn.cursor() as cur:
                    venda_row = cur.execute(
                        """
                        INSERT INTO venda (data_hora, nif_cliente)
                        VALUES (CURRENT_TIMESTAMP, %(nif_cliente)s)
                        RETURNING no_venda;
                        """,
                        {"nif_cliente": nif_cliente},
                    ).fetchone()

                    no_venda = venda_row.no_venda

                    for bilhete_data in bilhetes_preparados:
                        zonas = bilhete_data["zonas"]
                        desconto = bilhete_data["desconto"]

                        zonas_db = cur.execute(
                            """
                            SELECT id_zona, preco
                            FROM zona
                            WHERE id_zona = ANY(%(zonas)s);
                            """,
                            {"zonas": zonas},
                        ).fetchall()

                        precos_por_zona = {
                            row.id_zona: row.preco
                            for row in zonas_db
                        }

                        if set(precos_por_zona.keys()) != set(zonas):
                            zonas_invalidas = set(zonas) - set(precos_por_zona.keys())
                            raise ValueError(
                                f"Existem zonas inválidas: {sorted(zonas_invalidas)}."
                            )

                        bilhete_row = cur.execute(
                            """
                            INSERT INTO bilhete (desconto, no_venda)
                            VALUES (%(desconto)s, %(no_venda)s)
                            RETURNING bid;
                            """,
                            {
                                "desconto": desconto,
                                "no_venda": no_venda,
                            },
                        ).fetchone()

                        bid = bilhete_row.bid

                        for zona in zonas:
                            cur.execute(
                                """
                                INSERT INTO acesso (bid, id_zona)
                                VALUES (%(bid)s, %(zona)s);
                                """,
                                {
                                    "bid": bid,
                                    "zona": zona,
                                },
                            )

                        preco_base = sum(
                            Decimal(str(precos_por_zona[zona]))
                            for zona in zonas
                        )

                        preco_bilhete = (
                            preco_base * (Decimal("1.00") - desconto)
                        ).quantize(Decimal("0.01"))

                        preco_total += preco_bilhete

                        resultado_bilhetes.append(
                            {
                                "bid": bid,
                                "preco": float(preco_bilhete),
                            }
                        )

    except ValueError as e:
        return jsonify({"erro": str(e)}), 400

    return jsonify(
        {
            "no_venda": no_venda,
            "preco_total": float(preco_total.quantize(Decimal("0.01"))),
            "bilhetes": resultado_bilhetes,
        }
    ), 201

```

## PART III – Análise de Dados

A resolução das alíneas que se seguem **tem de obedecer às seguintes restrições**:
- Só pode haver **um único comando exterior** (CREATE, ALTER, UPDATE ou SELECT) e, no caso de envolver o comando SELECT, **não pode haver mais do que 1 nível de encadeamento de SELECTs**.
- **Não pode** recorrer aos operadores LIMIT ou WITH, ou a qualquer operador não coberto nos materiais da disciplina.


#### 4. Engenharia de Dados [2 valores]

##### Critérios de Avaliação

* Correção das soluções
* Eficiência das soluções
* Simplicidade das soluções

1. Crie uma vista materializada para auxiliar a direção do zoo a analisar as vendas de bilhetes e receitas do zoo e suas zonas, com o seguinte esquema:

    *vendas_zoo(bid, id_zona, mes, dia_da_semana, data, receita)*

    Em que teremos, para cada bilhete (*bid*) e zona (*id_zona*) a que o bilhete teve acesso, a data de venda do bilhete (e atributos derivados mes e dia_da_semana) e a receita da zona com o bilhete (i.e., o preço base da zona modificado pelo desconto).

In [15]:
%%sql zoo

CREATE MATERIALIZED VIEW vendas_zoo AS
SELECT
    b.bid,
    a.id_zona,
    EXTRACT(MONTH FROM v.data_hora) AS mes,
    EXTRACT(ISODOW FROM v.data_hora) AS dia_da_semana,
    DATE(v.data_hora) AS data,
    z.preco * (1 - b.desconto) AS receita
FROM bilhete b
JOIN venda v
    ON v.no_venda = b.no_venda
JOIN acesso a
    ON a.bid = b.bid
JOIN zona z
    ON z.id_zona = a.id_zona;

++
||
++
++

2. Modifique a tabela recinto adicionando um campo rentabilidade de tipo REAL

In [16]:
%%sql zoo

ALTER TABLE recinto
ADD COLUMN rentabilidade REAL;


++
||
++
++

3. Actualize a tabela recinto com o valor da rentabilidade sendo obtido pela receita total da zona onde está o recinto multiplicada pela fração entre os votos do recinto e o total de votos na zona. Pode recorrer à vista votos_zoo para realizar esta actualização.

In [17]:
%%sql zoo

UPDATE recinto r
SET rentabilidade =
(
    SELECT SUM(vz.receita)
    FROM vendas_zoo vz
    WHERE vz.id_zona = r.id_zona
)
*
(
    CAST(r.votos AS REAL)
    /
    (
        SELECT SUM(r2.votos)
        FROM recinto r2
        WHERE r2.id_zona = r.id_zona
    )
);

++
||
++
++

#### 5. Consultas [4 valores]

##### Critérios de Avaliação

* Correção das soluções
* Eficiência das soluções
* Simplicidade das soluções

Apresente a consulta SQL mais sucinta para cada um dos seguintes objetivos analíticos da empresa. Deve usar a vista vendas_zoo (exclusivamente ou em adição às outras tabelas) sempre que isso simplificar a consulta. Pode também usar operadores OLAP onde for adequado ao pedido.

1. Determinar o recinto mais rentável para cada zona, incluindo o valor da sua rentabilidade.

In [18]:
%%sql zoo

SELECT
    id_zona,
    id_recinto,
    rentabilidade
FROM (
    SELECT
        id_zona,
        id_recinto,
        rentabilidade,
        RANK() OVER (
            PARTITION BY id_zona
            ORDER BY rentabilidade DESC
        ) AS pos
    FROM recinto
) t
WHERE pos = 1;


id_zona,id_recinto,rentabilidade
1,1,208212.4
1,2,208212.4
1,3,208212.4
1,4,208212.4
1,5,208212.4
1,6,208212.4
1,7,208212.4
1,8,208212.4
1,9,208212.4
1,10,208212.4


2. Determinar se a aposta do zoo na sua especialidade está a compensar em termos de receitas, bilhetes ou votos, calculando as médias por zona, entre zonas da especialidade e zonas que não são da especialidade, de receitas globais, de bilhetes vendidos, e de votos recebidos.

In [19]:
%%sql zoo

SELECT
    CASE
        WHEN z.categoria IS NULL
            THEN 'Nao Especializada'
        ELSE 'Especializada'
    END AS tipo_zona,

    AVG(v.receita) AS media_receita,
    COUNT(DISTINCT v.bid) AS total_bilhetes,
    SUM(r.votos) AS total_votos

FROM zona z
LEFT JOIN vendas_zoo v
    ON z.id_zona = v.id_zona
LEFT JOIN recinto r
    ON z.id_zona = r.id_zona

GROUP BY
    CASE
        WHEN z.categoria IS NULL
            THEN 'Nao Especializada'
        ELSE 'Especializada'
    END;


tipo_zona,media_receita,total_bilhetes,total_votos
Especializada,14.1084898465703971,300000,22160020
Nao Especializada,8.2376274509803922,300000,6120000


3. Determinar a percentagem de bilhetes com acesso a todas as zonas, globalmente e com *drill downs* independentes por dia da semana e por mês.

In [20]:
%%sql zoo

SELECT
    mes,
    dia_da_semana,

    ROUND(
        100.0 *
        COUNT(
            DISTINCT CASE
                WHEN nzonas = 7 THEN bid
            END
        )
        /
        COUNT(DISTINCT bid),
        2
    ) AS perc_acesso_total

FROM (
    SELECT
        vz.bid,
        vz.mes,
        vz.dia_da_semana,
        COUNT(*) AS nzonas
    FROM vendas_zoo vz
    GROUP BY
        vz.bid,
        vz.mes,
        vz.dia_da_semana
) t

GROUP BY GROUPING SETS (
    (),
    (mes),
    (dia_da_semana)
)

ORDER BY mes, dia_da_semana;


mes,dia_da_semana,perc_acesso_total
1,None,2.00
2,None,2.00
3,None,2.00
4,None,2.00
5,None,2.00
6,None,2.00
None,1,2.00
None,2,2.00
None,3,2.00
None,4,2.00


4. Determinar que zonas podem ter um aumento no preço e que zonas devem ter uma redução no preço, analisando a média diária de bilhetes vendidos, globalmente e com vértices nas dimensões zona e mês.

In [21]:
%%sql zoo

SELECT
    id_zona,
    mes,

    ROUND(
        COUNT(DISTINCT bid)::NUMERIC / COUNT(DISTINCT data),
        2
    ) AS media_diaria_bilhetes

FROM vendas_zoo

GROUP BY CUBE(id_zona, mes)

ORDER BY id_zona, mes;


id_zona,mes,media_diaria_bilhetes
1,1,1870.97
1,2,1857.14
1,3,1870.97
1,4,1800.00
1,5,1967.74
1,6,1545.45
1,None,1851.85
2,1,1870.97
2,2,1857.14
2,3,1870.97


#### 6. Índices [3 valores]

##### Critérios de Avaliação

* Utilidade dos índices
* Adequação  da justificação

É expectável que seja necessário executar consultas semelhantes **às consultas 1 e 2 do ponto anterior** diversas vezes ao longo do tempo, pelo que pretendemos otimizar o desempenho dessas consultas por meio da criação de índices. Crie sobre a vista vendas\_zoo ou sobre as restantes tabelas da base de dados o(s) índice(s) que achar mais indicados para fazer essa otimização, justificando a sua escolha com **argumentos teóricos** e com **demonstração prática** do ganho em eficiência do índice por meio do comando EXPLAIN ANALYSE. Deve procurar uma otimização coletiva das duas consultas, evitando criar índices excessivos, particularmente se estes trazem apenas ganhos incrementais. Nomeadamente, devido a preocupações com o armazenamento, não se considera vantajoso atingir planos de execução com INDEX ONLY SCAN se isso obriga a colocar no índice mais atributos do que estritamente necessário para conseguir um INDEX SCAN.

##### Consulta 1

1. Demonstração do custo (com EXPLAIN ANALYSE)

In [22]:
%%sql zoo

DROP INDEX IF EXISTS idx_recinto_zona_rentabilidade;

EXPLAIN ANALYZE
SELECT
    id_zona,
    id_recinto,
    rentabilidade
FROM (
    SELECT
        id_zona,
        id_recinto,
        rentabilidade,
        RANK() OVER (
            PARTITION BY id_zona
            ORDER BY rentabilidade DESC
        ) AS pos
    FROM recinto
) AS t
WHERE pos = 1;

QUERY PLAN
Subquery Scan on t (cost=128.91..189.02 rows=9 width=12) (actual time=0.066..0.119 rows=160.00 loops=1)
Filter: (t.pos = 1)
Buffers: shared hit=2
-> WindowAgg (cost=128.91..165.89 rows=1850 width=20) (actual time=0.065..0.106 rows=160.00 loops=1)
Window: w1 AS (PARTITION BY recinto.id_zona ORDER BY recinto.rentabilidade ROWS UNBOUNDED PRECEDING)
Run Condition: (rank() OVER w1 <= 1)
Storage: Memory Maximum Storage: 17kB
Buffers: shared hit=2
-> Sort (cost=128.89..133.52 rows=1850 width=12) (actual time=0.057..0.062 rows=160.00 loops=1)
"Sort Key: recinto.id_zona, recinto.rentabilidade DESC"


2. Comandos de criação do(s) índice(s)

In [23]:
%%sql zoo

CREATE INDEX IF NOT EXISTS idx_recinto_zona_rentabilidade
ON recinto USING btree (id_zona, rentabilidade DESC);

ANALYZE recinto;

++
||
++
++

3. Demonstração do custo final (com EXPLAIN ANALYSE)

In [24]:
%%sql zoo

EXPLAIN ANALYZE
SELECT
    id_zona,
    id_recinto,
    rentabilidade
FROM (
    SELECT
        id_zona,
        id_recinto,
        rentabilidade,
        RANK() OVER (
            PARTITION BY id_zona
            ORDER BY rentabilidade DESC
        ) AS pos
    FROM recinto
) AS t
WHERE pos = 1;

QUERY PLAN
Subquery Scan on t (cost=9.48..14.66 rows=1 width=12) (actual time=0.038..0.163 rows=160.00 loops=1)
Filter: (t.pos = 1)
Buffers: shared hit=2
-> WindowAgg (cost=9.48..12.66 rows=160 width=20) (actual time=0.037..0.147 rows=160.00 loops=1)
Window: w1 AS (PARTITION BY recinto.id_zona ORDER BY recinto.rentabilidade ROWS UNBOUNDED PRECEDING)
Run Condition: (rank() OVER w1 <= 1)
Storage: Memory Maximum Storage: 17kB
Buffers: shared hit=2
-> Sort (cost=9.46..9.86 rows=160 width=12) (actual time=0.032..0.039 rows=160.00 loops=1)
"Sort Key: recinto.id_zona, recinto.rentabilidade DESC"


4. Justificação

Para a Consulta 1, foi criado um índice B-tree sobre recinto(id_zona, rentabilidade DESC), porque a consulta calcula o recinto mais rentável por zona através de uma função de janela com PARTITION BY id_zona e ORDER BY rentabilidade DESC. Teoricamente, este índice é adequado porque organiza os registos pela mesma ordem usada pela consulta. No entanto, o PostgreSQL continua a escolher Seq Scan, uma vez que a tabela recinto tem apenas 160 registos e a consulta precisa de analisar todos os recintos para calcular o ranking por zona. Assim, o custo de percorrer sequencialmente a tabela e ordenar em memória é inferior ao custo de usar o índice. Além disso, dentro dos limites do enunciado, cada zona só pode ter entre 10 e 30 recintos, pelo que a tabela continuará pequena e não é expectável obter naturalmente um Index Scan.

##### Consulta 2

1. Demonstração do custo inicial (com EXPLAIN ANALYSE)

In [25]:
%%sql zoo

DROP INDEX IF EXISTS idx_zona_vendas_zoo;

EXPLAIN ANALYZE
SELECT
    CASE
        WHEN z.categoria IS NULL
            THEN 'Nao Especializada'
        ELSE 'Especializada'
    END AS tipo_zona,

    AVG(v.receita) AS media_receita,
    COUNT(DISTINCT v.bid) AS total_bilhetes,
    SUM(r.votos) AS total_votos

FROM zona z
LEFT JOIN vendas_zoo v
    ON z.id_zona = v.id_zona
LEFT JOIN recinto r
    ON z.id_zona = r.id_zona

GROUP BY
    CASE
        WHEN z.categoria IS NULL
            THEN 'Nao Especializada'
        ELSE 'Especializada'
    END;

QUERY PLAN
GroupAggregate (cost=111301.45..119127.95 rows=200 width=80) (actual time=15537.142..17070.073 rows=2.00 loops=1)
Group Key: (CASE WHEN (z.categoria IS NULL) THEN 'Nao Especializada'::text ELSE 'Especializada'::text END)
"Buffers: shared hit=10435, temp read=277686 written=277939"
-> Sort (cost=111301.45..112866.25 rows=625920 width=72) (actual time=10000.866..13738.410 rows=28280020.00 loops=1)
"Sort Key: (CASE WHEN (z.categoria IS NULL) THEN 'Nao Especializada'::text ELSE 'Especializada'::text END), v.bid"
Sort Method: external merge Disk: 1110784kB
"Buffers: shared hit=10435, temp read=277686 written=277939"
-> Hash Right Join (cost=68.97..25366.57 rows=625920 width=72) (actual time=7.806..2035.056 rows=28280020.00 loops=1)
Hash Cond: (v.id_zona = z.id_zona)
Buffers: shared hit=10435


2. Comandos de criação do(s) índice(s)

In [26]:
%%sql zoo

CREATE INDEX IF NOT EXISTS idx_zona_vendas_zoo
ON vendas_zoo USING btree (id_zona);

++
||
++
++

3. Demonstração do custo final (com EXPLAIN ANALYSE)

In [27]:
%%sql zoo

EXPLAIN ANALYZE
SELECT
    CASE
        WHEN z.categoria IS NULL
            THEN 'Nao Especializada'
        ELSE 'Especializada'
    END AS tipo_zona,

    AVG(v.receita) AS media_receita,
    COUNT(DISTINCT v.bid) AS total_bilhetes,
    SUM(r.votos) AS total_votos

FROM zona z
LEFT JOIN vendas_zoo v
    ON z.id_zona = v.id_zona
LEFT JOIN recinto r
    ON z.id_zona = r.id_zona

GROUP BY
    CASE
        WHEN z.categoria IS NULL
            THEN 'Nao Especializada'
        ELSE 'Especializada'
    END;

QUERY PLAN
GroupAggregate (cost=246531.62..264209.12 rows=200 width=80) (actual time=15639.532..17109.927 rows=2.00 loops=1)
Group Key: (CASE WHEN (z.categoria IS NULL) THEN 'Nao Especializada'::text ELSE 'Especializada'::text END)
"Buffers: shared hit=10435, temp read=277686 written=277939"
-> Sort (cost=246531.62..250066.62 rows=1414000 width=72) (actual time=10222.075..13663.219 rows=28280020.00 loops=1)
"Sort Key: (CASE WHEN (z.categoria IS NULL) THEN 'Nao Especializada'::text ELSE 'Especializada'::text END), v.bid"
Sort Method: external merge Disk: 1110784kB
"Buffers: shared hit=10435, temp read=277686 written=277939"
-> Hash Right Join (cost=68.97..44083.47 rows=1414000 width=72) (actual time=16.315..2601.908 rows=28280020.00 loops=1)
Hash Cond: (v.id_zona = z.id_zona)
Buffers: shared hit=10435


4. Justificação

Foi criado um índice B-tree sobre vendas_zoo(id_zona), uma vez que a Consulta 2 junta a vista vendas_zoo com a tabela zona através do atributo id_zona. Assim, o índice escolhido incide diretamente sobre a coluna usada na condição de junção. Não foram incluídos outros atributos no índice, como bid ou receita, porque estes não são usados como filtros seletivos. O atributo bid é usado apenas no COUNT(DISTINCT) e receita é usada apenas na agregação AVG. Incluir estes atributos teria como objetivo aproximar a consulta de um Index Only Scan, mas isso aumentaria o armazenamento do índice e não seria adequado face ao enunciado. Apesar de o índice ser teoricamente adequado, o EXPLAIN ANALYZE pode continuar a apresentar Seq Scan sobre vendas_zoo. Isto acontece porque a consulta não tem WHERE seletivo e precisa de processar praticamente todos os registos para calcular as agregações por tipo de zona. Além disso, as tabelas envolvidas têm poucos registos, pelo que o otimizador pode considerar mais barato fazer um seq scan do que usar o índice. Assim, o índice sobre id_zona é uma escolha simples e justificada para a junção, mas o ganho prático pode ser reduzido no conjunto de dados atual.

#### Entrega

A submissão do projeto deve ser feita na forma de um arquivo zip de nome **entrega-bd-02-GG.zip**, onde **GG** é o número do grupo, estruturado da seguinte forma:

| Ficheiro | Descrição |
| :---- | :---- |
| entrega-bd-02-GG.ipynb | Um ficheiro Jupyter Notebook correspondendo ao preenchimento do template “entrega-bd-02-GG.ipynb” disponibilizado na página da disciplina (onde GG deverá ser subtituído pelo número do grupo).<br/><br/>Deverá preencher a primeira célula com o número do grupo, o número e nome dos alunos que o constituem, tal como a percentagem relativa de contribuição de cada aluno com o respectivo esforço (horas).<br/><br/>Deverá ainda preencher no Jupyter Notebook as respostas a todas as perguntas para as quais há um campo de resposta.<br/><br/>O ficheiro “entrega-bd-02-GG.ipynb” pode ser importado para o ambiente de trabalho disponibilizado para as aulas de laboratório (i.e., https://github.com/bdist/bdist-workspace), que serve de ambiente de teste para as partes em SQL.<br/><br/>Deve certificar-se que todo o código SQL é executável no ambiente de trabalho, |
| app/ | Pasta com os ficheiros que compõem a aplicação web correspondendo à secção<br/><br/>3\. **Desenvolvimento de Aplicação** |

**IMPORTANTE**

* Serão aplicadas penalizações aos grupos que não cumprirem o formato de submissão.
* Não serão aceites submissões fora do prazo.